legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageServer.java

```diff
@@ -187,6 +189,48 @@ public CompletableFuture<InitializeResult> initialize(InitializeParams initializ
         return initFuture;
     }
 
+    public <T> T runAndFireEvent(String eventType, Supplier<T> action)
+    {
+        Instant start = Instant.now();
+        Exception exception = null;
+        try
+        {
+            return action.get();
+        }
+        catch (Exception e)
+        {
+            exception = e;
+            throw e;
+        }
+        finally
+        {
+            this.fireEvent(eventType, start, Collections.emptyMap(), exception);
+        }
+    }
+
+    public void fireEvent(String eventType, Instant start, Map<String, Object> metadata, Throwable throwable)
+    {
+        JsonElement metadataAsTree = this.gson.toJsonTree(metadata);
+        Map<String, Object> finalMetadata = (Map<String, Object>) this.gson.fromJson(metadataAsTree, TypeToken.getParameterized(Map.class, String.class, Object.class));
+        if (throwable != null)
+        {
+            finalMetadata.put("error", true);
+            finalMetadata.put("errorMessage", throwable.getMessage());
+        }
+
+        this.globalState.findFeatureThatImplements(LegendUsageEventConsumer.class).forEach(x ->
+        {
+            try
+            {
+                x.consume(LegendUsageEventConsumer.event(eventType, start, Instant.now(), finalMetadata));
```

Comment:
I would say we should get the end timestamp just once, rather than creating a new one for each consumer.

Reply:
Yes - make sense.  I will create the event object outside too.